# 🚀딥러닝 모델 생성

# 📌모델 직접 개발하는 경우

## 1. 데이터 가져오기

In [1]:
# 데이터 준비
import tensorflow as tf

mnist = tf.keras.datasets.mnist
print(type(mnist))

<class 'module'>


In [2]:
# 훈련데이터, 검증데이터 가져오기
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# 확인
print("X_train:", X_train.shape, "\t X_test:", X_test.shape)
print("y_train:", y_train.shape, "\t y_test:", y_test.shape)

11501568/11490434 [==============================] - 0s 0us/step
X_train: (60000, 28, 28) 	 X_test: (10000, 28, 28)
y_train: (60000,) 	 y_test: (10000,)


## 2. 데이터 전처리

In [3]:
# 데이터 정규화 - 데이터의 범위를 일정하게 만듦
# 원본 데이터 값의 범위는 0-255
# 이를 0.0-1.0 사이로 정규화함
# 정규화를 할 때 분류 정확도가 개선되는 편
X_train, X_test = X_train/255.0, X_test/255.0

## 3. 모델 생성

In [24]:
# DNN
mlp_model = tf.keras.models.Sequential([
        tf.keras.layers.Flatten(input_shape = (28,28)),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dense(10, activation='softmax')
])

## 4. 모델 컴파일
### 손실함수 loss
- 회귀는 MSE(값 기준), 분류는 CrossEntropy(확률 기준)를 주로 이용함
- 회귀 : MSE, MAE, RMSE
- 이진분류 : binary_crossentropy
- 다중분류 : sparse_categorical_crossentropy 또는 categorical_crossentropy
- 원핫인코딩이 적용된 경우 : categorical_crossentropy
- 원핫인코딩이 적용되지 않은 경우 : sparse_categorical_crossentropy

In [ ]:
# 원핫인코딩
y_train = tf.keras.utils.to_categorical(y_train)
y_test = tf.keras.utils.to_categorical(y_test)

In [25]:
# 컴파일
# 이 경우 loss 함수는 categorical_crossentropy
mlp_model.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])
# sparse_categorical_crossentropy는 입력이 일반 정수임
# 각 샘플이 여러 개의 class에 속할 수 있거나, label이 soft probability일 때 사용함

## 5. 모델 확인

In [26]:
mlp_model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_2 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_4 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_5 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________


## 6. 모델 학습

In [27]:
mlp_model.fit(X_train, y_train, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.2647 - accuracy: 0.9247
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1153 - accuracy: 0.9657
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0785 - accuracy: 0.9765
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0598 - accuracy: 0.9825
Epoch 5/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0452 - accuracy: 0.9860
Epoch 6/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0366 - accuracy: 0.9883
Epoch 7/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0287 - accuracy: 0.9912
Epoch 8/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0223 - accuracy: 0.9933
Epoch 9/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0197 - accuracy: 0.9940
Epoch 10/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.015

## 7. 모델 평가

In [28]:
mlp_model.evaluate(X_test, y_test, verbose=1)

313/313 [==============================] - 1s 1ms/step - loss: 0.0756 - accuracy: 0.9800


[0.0756344422698021, 0.9800000190734863]

In [29]:
mlp_model.evaluate(X_test, y_test, verbose=2)

313/313 - 0s - loss: 0.0756 - accuracy: 0.9800


[0.0756344422698021, 0.9800000190734863]

# 📌함수를 이용한 모델 생성
- Functional API 이용

In [38]:
# DNN
mlp_model = tf.keras.models.Sequential([
        tf.keras.layers.Flatten(input_shape = (28,28)),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dense(10, activation='softmax')
])

input = tf.keras.Input(shape=(28,28))
x = tf.keras.layers.Flatten()(input)
x = tf.keras.layers.Dense(128, activation='relu')(x)
x = tf.keras.layers.Dense(10, activation='softmax')(x)
output = tf.keras.layers.Dense(10, activation='softmax')(x)

In [39]:
# 원핫인코딩
y_train = tf.keras.utils.to_categorical(y_train)
y_test = tf.keras.utils.to_categorical(y_test)

# 이 경우 loss 함수는 categorical_crossentropy
mlp_model.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])
# sparse_categorical_crossentropy는 입력이 일반 정수임
# 각 샘플이 여러 개의 class에 속할 수 있거나, label이 soft probability일 때 사용함

In [40]:
mlp_model.fit(X_train, y_train, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2623 - accuracy: 0.9247
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1152 - accuracy: 0.9656
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0780 - accuracy: 0.9769
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0595 - accuracy: 0.9815
Epoch 5/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0442 - accuracy: 0.9863
Epoch 6/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0350 - accuracy: 0.9890
Epoch 7/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0266 - accuracy: 0.9926
Epoch 8/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0237 - accuracy: 0.9925
Epoch 9/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0188 - accuracy: 0.9941
Epoch 10/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.014

# 📌상속을 이용한 모델 생성
- 모델의 재활용이 쉬움
- 모델 배포시 이용하는 방식

### ⚠ '_TupleWrapper' object is not callable
- 층을 쌓을 때 불필요한 쉼표(,)가 붙었는지 확인해보고 제거해주자

In [25]:
# 클래스 생성
class MLP_Model(tf.keras.Model):
    def __init__(self):
        super(MLP_Model, self).__init__()
        self.flatten = tf.keras.layers.Flatten()
        self.relu = tf.keras.layers.Dense(128, activation='relu')
        self.softmax = tf.keras.layers.Dense(10, activation='softmax')

    def call(self, inputs):
        x = self.flatten(inputs)
        x = self.relu(x)
        return self.softmax(x)

# 모델 생성
mlp_model = MLP_Model()

In [26]:
# 모델 컴파일
mlp_model.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])
# sparse_categorical_crossentropy는 입력이 일반 정수임
# 각 샘플이 여러 개의 class에 속할 수 있거나, label이 soft probability일 때 사용함

In [27]:
mlp_model.fit(X_train, y_train, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.2587 - accuracy: 0.9256
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1130 - accuracy: 0.9670
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0778 - accuracy: 0.9769
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0580 - accuracy: 0.9817
Epoch 5/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0448 - accuracy: 0.9865
Epoch 6/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0347 - accuracy: 0.9890
Epoch 7/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0281 - accuracy: 0.9913
Epoch 8/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0235 - accuracy: 0.9927
Epoch 9/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0195 - accuracy: 0.9941
Epoch 10/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.015

# 🚀 합성곱 신경망 이용

## 1. 데이터 가져오기

In [4]:
# 데이터 준비
import tensorflow as tf

mnist = tf.keras.datasets.mnist
print(type(mnist))

<class 'module'>


In [5]:
# 훈련데이터, 검증데이터 가져오기
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# 확인
print("X_train:", X_train.shape, "\t X_test:", X_test.shape)
print("y_train:", y_train.shape, "\t y_test:", y_test.shape)

X_train: (60000, 28, 28) 	 X_test: (10000, 28, 28)
y_train: (60000,) 	 y_test: (10000,)


In [6]:
# 데이터 정규화
X_train, X_test = X_train/255.0, X_test/255.0

## 📌2. 데이터 전처리 - CNN에 맞게 데이터 수정
- 다중 퍼셉트론 입력 이미지 : 높이, 너비, 2차원 텐서
- 합성곱 퍼셉트론 입력 이미지 : 높이, 너비, 3차원 텐서

In [11]:
# 입력데이터를 CNN에 맞게 수정
X_train_4d = X_train.reshape(-1, 28, 28, 1)
X_test_4d = X_test.reshape(-1, 28, 28, 1)

## 3. 모델링

In [7]:
# 모델 생성
cnn_model = tf.keras.models.Sequential([
            tf.keras.layers.Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1)),
            tf.keras.layers.MaxPool2D((2,2)),

            tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
            tf.keras.layers.MaxPool2D((2,2)),

            tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
            tf.keras.layers.Flatten(),

            tf.keras.layers.Dense(64, activation='relu'),
            tf.keras.layers.Dense(10, activation='softmax')
])

In [14]:
# 컴파일
# 이 경우 loss 함수는 categorical_crossentropy
cnn_model.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])

In [15]:
# 모델 확인
cnn_model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten (Flatten)            (None, 576)               0         
_________________________________________________________________
dense (Dense)                (None, 64)                3

## 4. 모델 훈련

In [16]:
# 훈련
cnn_model.fit(X_train_4d, y_train, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 42s 6ms/step - loss: 0.1534 - accuracy: 0.9540
Epoch 2/10
1875/1875 [==============================] - 11s 6ms/step - loss: 0.0459 - accuracy: 0.9857
Epoch 3/10
1875/1875 [==============================] - 11s 6ms/step - loss: 0.0330 - accuracy: 0.9895
Epoch 4/10
1875/1875 [==============================] - 11s 6ms/step - loss: 0.0250 - accuracy: 0.9920
Epoch 5/10
1875/1875 [==============================] - 11s 6ms/step - loss: 0.0203 - accuracy: 0.9934
Epoch 6/10
1875/1875 [==============================] - 11s 6ms/step - loss: 0.0158 - accuracy: 0.9950
Epoch 7/10
1875/1875 [==============================] - 11s 6ms/step - loss: 0.0131 - accuracy: 0.9960
Epoch 8/10
1875/1875 [==============================] - 11s 6ms/step - loss: 0.0105 - accuracy: 0.9967
Epoch 9/10
1875/1875 [==============================] - 11s 6ms/step - loss: 0.0098 - accuracy: 0.9966
Epoch 10/10
1875/1875 [==============================] - 11s 6ms/step - l

## 5. 모델 평가
- dnn 모델 : accuracy: 0.9800
- cnn 모델 : accuracy: 0.9911
- 101.1% 향상됨

In [18]:
cnn_model.evaluate(X_test_4d, y_test, verbose=2)

313/313 - 1s - loss: 0.0334 - accuracy: 0.9911


[0.033388931304216385, 0.991100013256073]

# 🚀 Keras API 모델 이용

## 1. 데이터 가져오기

In [1]:
# 데이터 준비
import tensorflow as tf

mnist = tf.keras.datasets.mnist
print(type(mnist))

# 훈련데이터, 검증데이터 가져오기
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# 확인
print("X_train:", X_train.shape, "\t X_test:", X_test.shape)
print("y_train:", y_train.shape, "\t y_test:", y_test.shape)

# 데이터 정규화
X_train, X_test = X_train/255.0, X_test/255.0

<class 'module'>
11501568/11490434 [==============================] - 0s 0us/step
X_train: (60000, 28, 28) 	 X_test: (10000, 28, 28)
y_train: (60000,) 	 y_test: (10000,)


## 📌2. 데이터 전처리 - ResNet에 맞게 데이터 수정
- CNN 모델이므로 차원을 하나 늘려주고, 이미지 사이즈를 조정함

In [2]:
# CNN에 맞게 입력데이터 수정
X_train_4d = X_train.reshape(-1, 28, 28, 1)
X_test_4d = X_test.reshape(-1, 28, 28, 1)

In [3]:
# ResNet에 맞게 입력데이터 수정
resized_X_train = tf.image.resize(X_train_4d, [32,32])
resized_X_test = tf.image.resize(X_test_4d, [32,32])

In [4]:
print(resized_X_train.shape)
print(resized_X_test.shape)

(60000, 32, 32, 1)
(10000, 32, 32, 1)


## 3. 모델링

In [5]:
# 모델 생성
resnet_model = tf.keras.applications.ResNet50V2(input_shape=(32,32,1),
                                                classes=10,
                                                weights=None)

In [6]:
# 컴파일
resnet_model.compile(optimizer='adam',
                     loss='sparse_categorical_crossentropy',
                     metrics=['accuracy'])

In [39]:
# resnet_model.summary()

## 4. 모델 훈련
- ResNet의 모델은 층 깊이가 깊어서 더 훈련하는 데 더 오랜 시간이 소요됨

In [7]:
# 훈련
resnet_model.fit(resized_X_train, y_train, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 220s 97ms/step - loss: 0.2323 - accuracy: 0.9382
Epoch 2/10
1875/1875 [==============================] - 180s 96ms/step - loss: 0.1192 - accuracy: 0.9694
Epoch 3/10
1875/1875 [==============================] - 180s 96ms/step - loss: 0.1106 - accuracy: 0.9730
Epoch 4/10
1875/1875 [==============================] - 179s 96ms/step - loss: 0.0810 - accuracy: 0.9787
Epoch 5/10
1875/1875 [==============================] - 186s 99ms/step - loss: 0.0693 - accuracy: 0.9821
Epoch 6/10
1875/1875 [==============================] - 182s 97ms/step - loss: 0.0593 - accuracy: 0.9844
Epoch 7/10
1875/1875 [==============================] - 180s 96ms/step - loss: 0.0523 - accuracy: 0.9863
Epoch 8/10
1875/1875 [==============================] - 180s 96ms/step - loss: 0.0385 - accuracy: 0.9889
Epoch 9/10
1875/1875 [==============================] - 180s 96ms/step - loss: 0.0360 - accuracy: 0.9901
Epoch 10/10
1875/1875 [==============================] 

In [12]:
# 평가
resnet_model.evaluate(resized_X_test, y_test, verbose=2)
# accuracy: 0.9897

313/313 - 10s - loss: 0.0364 - accuracy: 0.9897


[0.03640393167734146, 0.9897000193595886]

# 🚀 전이학습

## 1. 모델 찾기
- https://tfhub.dev/
    - mobilenet_v2_100_224
- 모델 이름 + feature_vector : 전이학습이 가능하도록 마지막 레이어(출력층)를 제거한 모델
- 모델 이름 + classification : 바로 분류가 가능한 전체 모델

### 2. 모델 가져오기
- url이 존재하면 Copy URL을 선택하여 URL 복사


In [ ]:
모델 이름 = tf.keras.Sequential([
	tensorflow_hun.KerasLayer(url, input_shape=(입력 구조),trainable=False),
	tf.keras.layers.Dense(출력의 개수)
	])

## 📌3. 모델 변환
- tensorflow lite 모델로 변환
- 확장자는 대부분의 경우 tflite

In [40]:
# cnn 모델을 변환하여 저장
# 변환기 생성
converter = tf.lite.TFLiteConverter.from_keras_model(cnn_model)

# 모델 변환 - convert 함수 호출
tflite_model = converter.convert()

# 변환된 모델을 파일에 기록함
with open('./keras_model.tflite', 'wb') as f:
    f.write(tflite_model)

INFO:tensorflow:Assets written to: /tmp/tmp3yh22p6g/assets


In [10]:
# resnet 모델을 변환하여 저장
# 변환기 생성
converter = tf.lite.TFLiteConverter.from_keras_model(resnet_model)

# 모델 변환 - convert 함수 호출
tflite_model = converter.convert()

# 변환된 모델을 파일에 기록함
with open('./keras_model_resnet.tflite', 'wb') as f:
    f.write(tflite_model)

INFO:tensorflow:Assets written to: /tmp/tmpjmo5npnr/assets


INFO:tensorflow:Assets written to: /tmp/tmpjmo5npnr/assets
/usr/local/lib/python3.7/dist-packages/keras/utils/generic_utils.py:497: CustomMaskWarning: Custom mask layers require a config and must override get_config. When loading, the custom mask layer must be passed to the custom_objects argument.
  category=CustomMaskWarning)


## 📌 4. 기기 배포
- 안드로이드 프로젝트의 assets 디렉토리에 .tflite 모델 복사

### Q. assets 디렉토리와 res 디렉토리의 차이?
- assets 디렉토리 : 내용을 필요할 때 읽어옴
- res 디렉토리 : 앱이 실행될 때 읽어옴. 주로 크기가 작은 데이터를 저장함.